source/agora/common/EnrollmentManager.d

```diff
@@ -0,0 +1,409 @@
+
+        Returns:
+            the number of validators in the validator set
+
+    ***************************************************************************/
+
+    public size_t getEnrollmentLength () @safe
+    {
+        return () @trusted {
+            return db.execute("SELECT count(*) FROM validator_set").oneValue!size_t;
+        }();
+    }
+
+    /***************************************************************************
+
+        Walk over the enrollment data in the set and call the provided delegate
+        with each hash and enrollment data
+
+        Params:
+            dg = the delegate to call
+
+        Returns:
+            the status code of the delegate, or zero
+
+    ***************************************************************************/
+
+    public int opApply (scope int delegate(const ref Hash hash, 
+        const ref Enrollment enroll) dg) @trusted
+    {
+        auto results = this.db.execute("SELECT key, val FROM validator_set");
+
+        foreach (row; results)
+        {
+            auto hash = *cast(Hash*)row.peek!(ubyte[])(0).ptr;
+            auto enroll = deserializeFull!Enrollment(row.peek!(ubyte[])(1));
+
+            if (auto ret = dg(hash, enroll))
+                return ret;
+        }
+
+        return 0;
+    }
+
+    /***************************************************************************
+
+        Remove the enrollment data with the given key from the validator set
+
+        Params:
+            enroll_hash = key for an enrollment data to remove
+
+    ***************************************************************************/
+
+    public void removeEnrollment (const ref Hash enroll_hash) @trusted
+    {
+        this.db.execute("DELETE FROM validator_set WHERE key = ?", enroll_hash[]);
+    }
+
+    /***************************************************************************
+
+        Make an enrollment data for enrollment process
+
+        Params:
+            frozen_utxo_hash = the hash of a frozen UTXO used to identify 
+                                a validator and to generate a siging key
+
+        Returns:
+            an enrollment data
+
+    ***************************************************************************/
+
+    public Enrollment createEnrollment (const ref Hash frozen_utxo_hash) @trusted
+    {
+        this.data = new Enrollment();
+
+        // K, frozen UTXO hash
+        this.data.utxo_key = frozen_utxo_hash;
+
+        // N, cycle length
+        this.data.cycle_length = 1008; // freezing period / 2
+
+        // generate random seed value
+        this.random_seed_src = Scalar.random();
+
+        // X, nth image of random seed
+        this.data.random_seed = hashFull(this.random_seed_src);
+        foreach (i; 0 .. this.data.cycle_length-1)
+            this.data.random_seed = hashFull(this.data.random_seed);
+
+        // R, signature noise
+        this.signature_noise = Pair.random();
+        this.data.R = this.signature_noise.V;
```

Comment:
@Geod24 I think I see an optimization opportunity here. We are storing `R` in both the `EnrollmentData` and the `(R, s)` signature pair. Could we omit R from the `EnrollmentData`? After all, `R` is already hashed in the challenge.